### ISSUE TITLE
YOLOE finetuning metrics

### ISSUE THREAD

**Issue by rogervendrell** [ultralytics/ultralytics#23774]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Rgearding YOLOE:
I have trained a fusion module that fuses tpe and vpe into a single representation. Training is done outside of YOLOE via contrastive learning, so I've just put the trained weights inside get_cls_pe()

```python
def get_cls_pe(self, tpe, vpe, batch, fusion_weights):
        """Get class positional embeddings.

        Args:
            tpe (torch.Tensor | None): Text positional embeddings.
            vpe (torch.Tensor | None): Visual positional embeddings.

        Returns:
            (torch.Tensor): Class positional embeddings.
        """
        if tpe is None or vpe is None:
            return self.get_cls_pe_original(tpe, vpe)
        
        if not hasattr(self, "fusion_module"):
            self.initialise_fusion(embed_dim=512)
        
        batch_index = batch["batch_idx"].long()
        class_index = batch["cls"].squeeze(-1).long()

        tpe2 = tpe.detach()
        vpe2 = vpe.detach()

        t_cls = tpe2[batch_index, class_index] # [N, D]
        v_cls = vpe2[batch_index, class_index] # [N, D]

        fused = self.fusion_module(
            tpe=t_cls,
            vpe=v_cls
        ) # [N, D]

        diff = torch.zeros_like(tpe)
        diff[batch_index, class_index] = fused - tpe2[batch_index, class_index]
        return tpe + diff
```

Now I am trying to finetune YOLOE to adapt to this new embedding space by freezing everything (the fusion module also) except the **cv3** module in the classification head (which, if I understood it correctly, projects prompt embeddings to object embeddings in the image). This is the training script:

```python
import torch
torch.autograd.set_detect_anomaly(True)
import os
from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOESegVPTrainer, YOLOEVPTrainer
import wandb
from logs import log_losses_to_wandb


PATH = "/LENA/datasets/"
os.environ["WANDB_PROJECT"] = "yoloe-seg-vp"
WANDB = False

data = dict(
    train=dict(
        yolo_data=["coco.yaml"],
    ),
    val=dict(yolo_data=["coco.yaml"]),
)


model = YOLOE("pretrained_weights/yoloe-11s-seg.pt")
# model = YOLOE("yoloe-11l-seg-det.pt")

# freeze every layer except for the savpe and reprta modules.
head_index = len(model.model.model) - 1
freeze = list(range(0, head_index))
for name, child in model.model.model[-1].named_children():
    if name not in ["cv3"]:
        freeze.append(f"{head_index}.{name}")

train_args = dict(
    batch=64,
    epochs=10,
    optimizer="AdamW",
    lr0=0.00001,
    warmup_bias_lr=0.1,
    weight_decay=0.0005,
    momentum=0.9,
    workers=0,
    device="0",
    freeze=freeze,
    save=True,
    amp=False,
    # Augmentations
    hsv_h=0.0,
    hsv_s=0.0,
    hsv_v=0.0,
    translate=0.0,
    scale=0.0,
    fliplr=0.0,
    mosaic=0.0,
    erasing=0.0,
    auto_augment=None,
)

if WANDB:
    wandb.init(
        project="yoloe-seg-vp",
        name="yoloe-11s-seg-savpe-reprta",
        config=train_args,
    )
    model.add_callback("on_fit_epoch_end", log_losses_to_wandb)

model.train(
    data=data,
    trainer=YOLOESegVPTrainer,
    **train_args,
 )
```

Training collapses to predicting background for every class. Only the "person" class gets some detections right. These are the normalised confusion matrix and an example of some predictions:

<img width="3000" height="2250" alt="Image" src="https://github.com/user-attachments/assets/92fd0a21-869c-4a36-a42e-5c26fcf3193e" />

![Image](https://github.com/user-attachments/assets/b4bcd898-17c3-4236-b787-c1bdbdb0f15a)

My two questions are:

1. Why are the reported metrics so high if the model is only predicting one class?
2. Why is it collapsing on just one class?

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/e5ec67d5-6449-409b-a12b-571a5c1c3b60" />

### Additional

Additionally, trying to understand why only "person" instances are being predicted, I've checked whether the fusion for "person" and an image of a person is somewhat different from the fusion of other embeddings, but nothing indicates so.

![Image](https://github.com/user-attachments/assets/4c360816-53ff-4c6b-93b7-90bd12501d48)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rogervendrell, thanks for the detailed report and for using Ultralytics 🚀 This is an automated response to help triage your issue quickly; an Ultralytics engineer will also follow up soon.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) for usage references, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.

Since this appears to be a 🐛 bug/behavior report (unexpected metric behavior and class collapse during YOLOE fine-tuning), please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that reproduces the issue with the smallest possible script/config so we can verify and debug efficiently 🔍

If this is part of a custom training ❓ workflow, please include as much detail as possible (exact model checkpoint, full training/val logs, dataset subset details, and any code modifications), and confirm alignment with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by rogervendrell** [comment]
This is the .zip of the training run:

[train65.zip](https://github.com/user-attachments/files/25679509/train65.zip)

**Comment by glenn-jocher** [comment]
Thanks for sharing the run zip — if the validator is running in prompt-based mode (or `single_cls=True` / `classes=...` is set anywhere), the reported mAP can be “high” because it’s computed over a restricted prompted subset rather than a full 80-class COCO-style evaluation; can you paste the text contents of your run’s `args.yaml` plus `!yolo checks`, and also try a plain COCO eval with:

```bash
yolo val model=/path/to/best.pt data=coco.yaml
```

On the collapse side, please first confirm you’re *actually* training only `cv3[*][2]` (it’s easy to accidentally freeze it with custom `freeze` lists); the built-in linear-probing trainers explicitly `deepcopy()` the `cv3[*][2]` convs to ensure they remain trainable after prompt fusion (see the [YOLOE segmentation linear-probing trainer `get_model()`](https://docs.ultralytics.com/reference/models/yolo/yoloe/train_seg/)), and also make sure your fused prompt vectors are L2-normalized (the head normalizes text prompts in [`YOLOEDetect.get_tpe()`](https://docs.ultralytics.com/reference/nn/modules/head/)) to avoid score saturation toward frequent classes like “person”.

```python
for n, p in model.model.named_parameters():
    if p.requires_grad:
        print(n, p.shape)
```

**Comment by rogervendrell** [comment]
This is `args.yaml` of the training run, neither `single_cls` nor `classes` is set. The only unfrozen module is 23.cv3, as expected.

```yaml
task: segment
mode: train
model: pretrained_weights/yoloe-11s-seg.pt
data:
  train:
    yolo_data:
    - coco.yaml
  val:
    yolo_data:
    - coco.yaml
epochs: 10
time: null
patience: 100
batch: 64
imgsz: 640
save: true
save_period: -1
cache: false
device: '0'
workers: 0
project: null
name: train65
exist_ok: false
pretrained: true
optimizer: AdamW
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: false
fraction: 1.0
profile: false
freeze:
- 0
- 1
- 2
- 3
- 4
- 5
- 6
- 7
- 8
- 9
- 10
- 11
- 12
- 13
- 14
- 15
- 16
- 17
- 18
- 19
- 20
- 21
- 22
- 23.cv2
- 23.dfl
- 23.cv4
- 23.proto
- 23.cv5
- 23.reprta
- 23.savpe
multi_scale: 0.0
compile: false
overlap_mask: false
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
end2end: null
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.0001
lrf: 0.01
momentum: 0.9
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
rle: 1.0
angle: 1.0
nbs: 64
hsv_h: 0.0
hsv_s: 0.0
hsv_v: 0.0
degrees: 0.0
translate: 0.0
scale: 0.0
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.0
bgr: 0.0
mosaic: 0.0
mixup: 0.0
cutmix: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: null
erasing: 0.0
cfg: null
tracker: botsort.yaml
save_dir: /home/rvendrell/Documents/GitHub/ultralytics/runs/segment/train65
```

Adding an additional `Normalize` before returning the fused embeddings in the forward pass of the fusion module (see below) did not solve this. Some runs happened to collapse on some other classes (it is not always the person class, but it always collapses on one class). Still, performance in that class is not good, it just happens to predict some instances and completely ignore the rest.

```python
#ultralytics/nn/modules/fusion.py
import torch
import torch.nn as nn
import torch.nn.functional as F


class Fusion(nn.Module):
    def __init__(self, embed_dim):
        super().__init__()

        # Normalize independently
        self.t_norm = nn.LayerNorm(embed_dim)
        self.v_norm = nn.LayerNorm(embed_dim)

        # Normalize concatenated
        self.concat_norm = nn.LayerNorm(2 * embed_dim)

        # Gating network
        self.gate_mlp = nn.Sequential(
            nn.Linear(2 * embed_dim, 4*embed_dim),  
            nn.GELU(),
            nn.LayerNorm(4*embed_dim),
            nn.Linear(4*embed_dim, embed_dim)
        )

        # Normalize gate logits before sigmoid
        self.gate_out_norm = nn.LayerNorm(embed_dim)

        # Initialise weights
        self.apply(self._init_weights)

    def _init_weights(self, module):
        if isinstance(module, nn.Linear):
            nn.init.xavier_uniform_(module.weight)
            if module.bias is not None:
                nn.init.zeros_(module.bias)
        elif isinstance(module, nn.LayerNorm):
            nn.init.ones_(module.weight)
            nn.init.zeros_(module.bias)

        nn.init.zeros_(self.gate_mlp[-1].weight)
        nn.init.zeros_(self.gate_mlp[-1].bias)

    def forward(self, tpe, vpe):
        """
        tpe: [N, embed_dim]
        vpe: [N, embed_dim]
        """
        if tpe is None and vpe is None:
            return None
        if tpe is None:
            return vpe
        if vpe is None:
            return tpe

        assert tpe.shape == vpe.shape, \
            f"got mismatching shapes for tpe {tpe.shape} and vpe {vpe.shape}"

        t = self.t_norm(tpe)
        v = self.v_norm(vpe)

        combined = torch.cat([t, v], dim=-1)
        combined = self.concat_norm(combined)

        gate_logits = self.gate_mlp(combined)
        gate_logits = self.gate_out_norm(gate_logits)

        g = torch.sigmoid(gate_logits)

        fused = g * tpe + (1.0 - g) * vpe
        return F.normalize(fused, p=2, dim=-1)
```

@Y-T-G @ShuaiLYU I'd appreciate it if any of you could take a look at this. Thank you very much

**Comment by ShuaiLYU** [comment]
Hi @rogervendrell,

**Q1 – Collapse to "person" + inflated metrics** 
This is likely a data issue. "person" is by far the most frequent class in coco. it seems that the model  finds it easiest shortcout when the embedding is fused. COCO mAP averages per-class AP accross all 80 classes;  a strong "person" AP can inflate the overall mAP, even other 70 classes are zero. 

**Q2 – Suggested ablations to isolate the problem**
Before starting the  fusion experiments,  it would be helpful to do some controlled comparison:
- Train with the **same setup but without the fusion module** (i.e. standard `get_cls_pe`). If results are significantly better, the fusion is introducing the issue.
-  train only cv3[*][2], as glenn mention above.

**Comment by rogervendrell** [comment]
Training only cv3[*][2] reduces the "collapsing" effect but predictions are still bad.

<img width="3000" height="2250" alt="Image" src="https://github.com/user-attachments/assets/a888c65b-798b-4d8b-967c-6843323b5787" />

> Train with the same setup but without the fusion module (i.e. standard get_cls_pe). If results are significantly better, the fusion is introducing the issue.

I am aware that introducing the fusion module is the issue here, what I'd need to find out is what's the proper way of allowing the classification head to adapt to these new fused embeddings.

**Comment by glenn-jocher** [comment]
That helps narrow it down — the supported `cv3`-only adaptation path assumes a *fixed* class embedding space. In the built-in linear-probing flow, [YOLOEPESegTrainer.get_model()](https://docs.ultralytics.com/reference/models/yolo/yoloe/train_seg/) does `set_classes(...)` and [`fuse(...)`](https://docs.ultralytics.com/reference/nn/modules/head/) once, then trains only `cv3[*][2]` (and `one2one_cv3[*][2]` if present), so if your fusion changes the class embedding per image or batch inside `get_cls_pe()`, `cv3` has no stationary target and collapse is expected. If you want to adapt only the classifier, precompute one fused embedding per class and load it with [`YOLOEModel.set_classes()`](https://docs.ultralytics.com/reference/nn/tasks/); if the fusion must stay sample-conditioned, you’ll likely need to keep the head unfused and unfreeze more than just `cv3`.
